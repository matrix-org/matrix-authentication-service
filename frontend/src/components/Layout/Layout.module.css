/* Copyright 2023 The Matrix.org Foundation C.I.C.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

.layout-container {
  box-sizing: border-box;
  display: flex;
  flex-direction: column;

  max-width: calc(420px + var(--cpd-space-5x) * 2);

  /* Fallback for browsers that do not support 100svh */
  min-height: 100vh;
  min-height: 100svh;

  margin: 0 auto;
  padding-inline: var(--cpd-space-5x);
  padding-block: var(--cpd-space-12x);
  gap: var(--cpd-space-8x);

  &.wide {
    max-width: calc(520px + var(--cpd-space-5x) * 2);
  }
}

@media screen and (min-width: 768px) {
  .layout-container {
    padding-block-start: var(--cpd-space-20x);
    padding-block-end: var(--cpd-space-10x);
  }
}
