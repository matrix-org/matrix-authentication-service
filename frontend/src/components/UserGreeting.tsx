// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { Heading, Text, Avatar } from "@vector-im/compound-web";

import { FragmentType, graphql, useFragment } from "../gql";

import UnverifiedEmailAlert from "./UnverifiedEmailAlert";
import styles from "./UserGreeting.module.css";

export const USER_GREETING_FRAGMENT = graphql(/* GraphQL */ `
  fragment UserGreeting_user on User {
    id
    username
    matrix {
      mxid
      displayName
    }

    ...UnverifiedEmailAlert
  }
`);

type Props = {
  user: FragmentType<typeof USER_GREETING_FRAGMENT>;
};

const UserGreeting: React.FC<Props> = ({ user }) => {
  const data = useFragment(USER_GREETING_FRAGMENT, user);

  return (
    <>
      <header className={styles.header}>
        <Avatar
          size="var(--cpd-space-24x)"
          id={data.matrix.mxid}
          name={data.matrix.displayName || data.matrix.mxid}
        />
        <Heading size="xl" weight="semibold">
          {data.matrix.displayName || data.username}
        </Heading>
        <Text size="lg" className={styles.mxid}>
          {data.matrix.mxid}
        </Text>
      </header>
      <UnverifiedEmailAlert user={data} />
    </>
  );
};

export default UserGreeting;
