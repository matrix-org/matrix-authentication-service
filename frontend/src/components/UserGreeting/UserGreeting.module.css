/* Copyright 2023 The Matrix.org Foundation C.I.C.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

.user {
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: var(--cpd-space-4x);
  outline: 1px solid var(--cpd-color-gray-400);
  outline-offset: -1px;
  border-radius: var(--cpd-space-3x);
  padding: var(--cpd-space-4x);
}

.meta {
  flex: 1 1;

  /* Make sure the text properly truncates */
  min-width: 0;

  & p {
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
  }
}

.edit-button {
  align-self: flex-start;
}

.mxid {
  color: var(--cpd-color-text-secondary);
}

.dialog-form {
  display: flex;
  flex-direction: column;
  gap: var(--cpd-space-4x);
  align-self: center;
  max-width: 378px;
  width: 100%;
  margin-block-end: var(--cpd-space-9x);
}
