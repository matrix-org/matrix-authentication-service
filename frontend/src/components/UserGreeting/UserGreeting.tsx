// Copyright 2023, 2024 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import IconClose from "@vector-im/compound-design-tokens/icons/close.svg?react";
import IconEdit from "@vector-im/compound-design-tokens/icons/edit.svg?react";
import {
  Text,
  Avatar,
  IconButton,
  Tooltip,
  Button,
  Form,
} from "@vector-im/compound-web";
import { ComponentPropsWithoutRef, forwardRef, useRef, useState } from "react";
import { useTranslation } from "react-i18next";
import { useMutation } from "urql";

import { FragmentType, graphql, useFragment } from "../../gql";
import { SetDisplayNameStatus } from "../../gql/graphql";
import * as Dialog from "../Dialog";
import LoadingSpinner from "../LoadingSpinner";

import styles from "./UserGreeting.module.css";

export const FRAGMENT = graphql(/* GraphQL */ `
  fragment UserGreeting_user on User {
    id
    matrix {
      mxid
      displayName
    }
  }
`);

export const CONFIG_FRAGMENT = graphql(/* GraphQL */ `
  fragment UserGreeting_siteConfig on SiteConfig {
    id
    displayNameChangeAllowed
  }
`);

const SET_DISPLAYNAME_MUTATION = graphql(/* GraphQL */ `
  mutation SetDisplayName($userId: ID!, $displayName: String) {
    setDisplayName(input: { userId: $userId, displayName: $displayName }) {
      status
      user {
        id
        matrix {
          displayName
        }
      }
    }
  }
`);

// This needs to be its own component because else props and refs aren't passed properly in the trigger
const EditButton = forwardRef<
  HTMLButtonElement,
  { label: string } & ComponentPropsWithoutRef<"button">
>(({ label, ...props }, ref) => (
  <Tooltip label={label}>
    <IconButton
      ref={ref}
      type="button"
      size="var(--cpd-space-6x)"
      className={styles.editButton}
      {...props}
    >
      <IconEdit />
    </IconButton>
  </Tooltip>
));

type Props = {
  user: FragmentType<typeof FRAGMENT>;
  siteConfig: FragmentType<typeof CONFIG_FRAGMENT>;
};

const UserGreeting: React.FC<Props> = ({ user, siteConfig }) => {
  const fieldRef = useRef<HTMLInputElement>(null);
  const data = useFragment(FRAGMENT, user);
  const { displayNameChangeAllowed } = useFragment(CONFIG_FRAGMENT, siteConfig);

  const [setDisplayNameResult, setDisplayName] = useMutation(
    SET_DISPLAYNAME_MUTATION,
  );

  const [open, setOpen] = useState(false);
  const { t } = useTranslation();

  const onSubmit = async (
    event: React.FormEvent<HTMLFormElement>,
  ): Promise<void> => {
    event.preventDefault();

    const form = event.currentTarget;
    const formData = new FormData(form);
    const displayName = (formData.get("displayname") as string) || null;

    const result = await setDisplayName({ displayName, userId: data.id });

    if (result.data?.setDisplayName.status === SetDisplayNameStatus.Set) {
      setOpen(false);
    }
  };

  return (
    <div className={styles.user}>
      <Avatar
        size="var(--cpd-space-14x)"
        id={data.matrix.mxid}
        name={data.matrix.displayName || data.matrix.mxid}
        className={styles.avatar}
      />
      <div className={styles.meta}>
        {data.matrix.displayName ? (
          <>
            <Text size="lg" weight="semibold">
              {data.matrix.displayName}
            </Text>
            <Text size="md" className={styles.mxid}>
              {data.matrix.mxid}
            </Text>
          </>
        ) : (
          <Text size="lg" weight="semibold">
            {data.matrix.mxid}
          </Text>
        )}
      </div>

      {displayNameChangeAllowed && (
        <Dialog.Dialog
          trigger={<EditButton label={t("action.edit")} />}
          open={open}
          onOpenChange={(open) => {
            // Reset the form when the dialog is opened or closed
            fieldRef.current?.form?.reset();
            setOpen(open);
          }}
        >
          <Dialog.Title>
            {t("frontend.account.edit_profile.title")}
          </Dialog.Title>

          <Avatar
            size="88px"
            className="self-center"
            id={data.matrix.mxid}
            name={data.matrix.displayName || data.matrix.mxid}
          />

          <Form.Root onSubmit={onSubmit}>
            <div className={styles.dialogForm}>
              <Form.Field
                name="displayname"
                serverInvalid={
                  setDisplayNameResult.data?.setDisplayName.status ===
                  SetDisplayNameStatus.Invalid
                }
              >
                <Form.Label>
                  {t("frontend.account.edit_profile.display_name_label")}
                </Form.Label>

                <Form.ActionControl
                  type="text"
                  Icon={IconClose}
                  autoComplete="name"
                  defaultValue={data.matrix.displayName || undefined}
                  actionLabel={t("action.clear")}
                  ref={fieldRef}
                  onActionClick={() => {
                    if (fieldRef.current) {
                      fieldRef.current.value = "";
                      fieldRef.current.focus();
                    }
                  }}
                />

                <Form.HelpMessage>
                  {t("frontend.account.edit_profile.display_name_help")}
                </Form.HelpMessage>
              </Form.Field>

              <Form.Field name="mxid">
                <Form.Label>
                  {t("frontend.account.edit_profile.username_label")}
                </Form.Label>
                <Form.TextControl value={data.matrix.mxid} readOnly />
              </Form.Field>
            </div>

            <Form.Submit disabled={setDisplayNameResult.fetching}>
              {setDisplayNameResult.fetching && <LoadingSpinner inline />}
              {t("action.save")}
            </Form.Submit>
          </Form.Root>

          <Dialog.Close asChild>
            <Button kind="tertiary">{t("action.cancel")}</Button>
          </Dialog.Close>
        </Dialog.Dialog>
      )}
    </div>
  );
};

export default UserGreeting;
