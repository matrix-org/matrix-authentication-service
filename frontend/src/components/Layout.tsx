// Copyright 2022 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { Link } from "@vector-im/compound-web";
import { useAtomValue } from "jotai";

import { currentUserIdAtom } from "../atoms";
import { isErr, unwrapErr, unwrapOk } from "../result";

import GraphQLError from "./GraphQLError";
import styles from "./Layout.module.css";
import NavBar from "./NavBar";
import NavItem from "./NavItem";
import NotLoggedIn from "./NotLoggedIn";
import UserGreeting from "./UserGreeting";

const Layout: React.FC<{ children?: React.ReactNode }> = ({ children }) => {
  const result = useAtomValue(currentUserIdAtom);
  if (isErr(result)) return <GraphQLError error={unwrapErr(result)} />;

  const userId = unwrapOk(result);
  if (userId === null)
    return (
      <div className={styles.container}>
        <NotLoggedIn />
      </div>
    );

  return (
    <div className={styles.container}>
      <UserGreeting userId={userId} />

      <NavBar>
        <NavItem route={{ type: "home" }}>Home</NavItem>
        <NavItem route={{ type: "email-list" }}>Emails</NavItem>
      </NavBar>

      <main>{children}</main>

      <footer className={styles.footer}>
        <nav className={styles.footerLinks}>
          <ul>
            <Link href="https://matrix.org/legal/copyright-notice">Info</Link>
            <Link href="https://matrix.org/legal/privacy-notice">Privacy</Link>
            <Link href="https://matrix.org/legal/terms-and-conditions">
              Terms & Conditions
            </Link>
          </ul>
        </nav>
      </footer>
    </div>
  );
};

export default Layout;
