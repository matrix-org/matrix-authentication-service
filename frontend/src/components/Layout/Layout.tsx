// Copyright 2022 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { useAtomValue } from "jotai";
import { useTranslation } from "react-i18next";

import { appConfigAtom } from "../../routing";
import Footer from "../Footer";
import NavBar from "../NavBar";
import NavItem from "../NavItem";
import UserGreeting from "../UserGreeting";

import styles from "./Layout.module.css";

const Layout: React.FC<{
  user: React.ComponentProps<typeof UserGreeting>["user"];
  children?: React.ReactNode;
}> = ({ user, children }) => {
  const appConfig = useAtomValue(appConfigAtom);
  const { t } = useTranslation();

  return (
    <div className={styles.layoutContainer}>
      <UserGreeting user={user} />

      <NavBar>
        <NavItem to="/">{t("frontend.nav.profile")}</NavItem>
        <NavItem to="/sessions">{t("frontend.nav.sessions")}</NavItem>
      </NavBar>

      {children}

      <Footer
        imprint={appConfig.branding?.imprint}
        tosUri={appConfig.branding?.tosUri}
        policyUri={appConfig.branding?.policyUri}
      />
    </div>
  );
};

export default Layout;
