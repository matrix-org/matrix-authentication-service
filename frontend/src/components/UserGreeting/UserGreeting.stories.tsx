// Copyright 2024 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import type { Meta, StoryObj } from "@storybook/react";
import { Provider } from "urql";
import { fromValue, delay, pipe } from "wonka";

import { makeFragmentData } from "../../gql";
import {
  SetDisplayNameMutation,
  SetDisplayNameStatus,
} from "../../gql/graphql";

import UserGreeting, { CONFIG_FRAGMENT, FRAGMENT } from "./UserGreeting";

const Template: React.FC<{
  displayName?: string;
  mxid: string;
  displayNameChangeAllowed: boolean;
}> = ({ displayName, mxid, displayNameChangeAllowed }) => {
  const userId = "user id";

  const mockClient = {
    /* This will resolve after a small delay */
    executeMutation: () =>
      pipe(
        fromValue({
          data: {
            setDisplayName: {
              status: SetDisplayNameStatus.Set,
              user: { id: userId, matrix: { displayName } },
            },
          },
        } satisfies { data: SetDisplayNameMutation }),
        delay(300),
      ),
  };

  const user = makeFragmentData(
    {
      id: "user id",
      matrix: {
        mxid,
        displayName,
      },
    },
    FRAGMENT,
  );

  const config = makeFragmentData(
    {
      id: "site config id",
      displayNameChangeAllowed,
    },
    CONFIG_FRAGMENT,
  );

  return (
    <Provider value={mockClient}>
      <UserGreeting user={user} siteConfig={config} />
    </Provider>
  );
};

const meta = {
  title: "UI/User Greeting",
  component: Template,
  args: {
    displayNameChangeAllowed: true,
    displayName: "Kilgore Trout",
    mxid: "@kilgore:matrix.org",
  },
  argTypes: {
    displayNameChangeAllowed: {
      control: "boolean",
    },
    displayName: {
      control: "text",
    },
    mxid: {
      control: "text",
    },
  },
} satisfies Meta<typeof Template>;

export default meta;
type Story = StoryObj<typeof Template>;

export const Basic: Story = {};

export const NoDisplayName: Story = {
  args: {
    displayName: undefined,
  },
};

export const DisplayNameChangeNotAllowed: Story = {
  args: {
    displayNameChangeAllowed: false,
  },
};
