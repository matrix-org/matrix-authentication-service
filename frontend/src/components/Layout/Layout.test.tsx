// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// @vitest-environment happy-dom

import { render } from "@testing-library/react";
import { Provider } from "urql";
import { describe, expect, it } from "vitest";
import { never } from "wonka";

import { makeFragmentData } from "../../gql";
import { DummyRouter } from "../../test-utils/router";
import { USER_GREETING_FRAGMENT } from "../UserGreeting";

import Layout from "./Layout";

describe("<Layout />", () => {
  it("renders app navigation correctly", async () => {
    const mockClient = {
      executeQuery: (): typeof never => never,
    };

    const user = makeFragmentData(
      {
        id: "abc123",
        username: "alice",
        matrix: {
          mxid: "@alice:example.org",
          displayName: "Alice",
        },
      },
      USER_GREETING_FRAGMENT,
    );

    const component = render(
      <Provider value={mockClient}>
        <DummyRouter>
          <Layout user={user} />
        </DummyRouter>
      </Provider>,
    );

    expect(await component.findByText("Profile")).toMatchSnapshot();
    expect(await component.findByText("Sessions")).toMatchSnapshot();
  });
});
