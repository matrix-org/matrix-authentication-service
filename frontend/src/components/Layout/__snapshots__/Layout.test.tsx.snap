// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Layout /> > renders app navigation correctly 1`] = `
<a
  aria-current="page"
  class="_navItem_8603fc"
  data-status="active"
  href="/"
>
  Profile
</a>
`;

exports[`<Layout /> > renders app navigation correctly 2`] = `
<a
  class="_navItem_8603fc"
  href="/sessions"
>
  Sessions
</a>
`;
